tracks/bash/exercises/hamming/mentoring.md

```diff
@@ -0,0 +1,127 @@
+### Reasonable Solutions
+
+```bash
+#!/usr/bin/env bash
+main () {
+  (( "$#" == 2 )) || {
+      echo "Usage: hamming.sh <string1> <string2>"
+      exit 1
+  }
+  
+  # Regular vars are easier to read when doing fancy parameter expansion.
+  a=$1 b=$2 
+  
+  (( "${#a}" == "${#b}" )) || {
+      echo "left and right strands must be of equal length"
+      exit 1
+  }
+
+  declare -i count
+  for (( i = 0; i < "${#a}"; i++ )); do
+    [[ "${a:i:1}" == "${b:i:1}" ]] || count+=1
+  done
+
+  printf '%d\n' "$count"
+}
+
+main "$@"
+```
+```bash
+#!/usr/bin/env bash
+main () {
+    if (( $# != 2 )); then 
+        echo "Usage: hamming.sh <string1> <string2>"
+        exit 1
+    fi
+    if (( ${#1} != ${#2} )); then
+        echo "left and right strands must be of equal length"
+        exit 1
+    fi
+    hamming=0
+    for (( i=0; i<${#1}; i++ )); do
+        [[ "${1:$i:1}" != "${2:$i:1}" ]] && (( ++hamming ))
+    done
+    echo "$hamming"
+}
+
+main "$@"
+```
+
+
+### Common Suggestions
+
+* Suggest using a for loop in bash or other [looping constructs](https://www.gnu.org/software/bash/manual/html_node/Looping-Constructs.html#Looping-Constructs) https://www.gnu.org/software/bash/manual/html_node/Looping-Constructs.html#Looping-Constructs
+
+* Suggest using [Parameter Expansion - Substring Extraction](https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html) to access the index of the string. https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html
+
+* Suggest creating a counter to gather hammingDistance
+1) let "hammingDistance+=1"
+2) counter=0; (( ++counter ));
+3) counter=0; counter=(( $counter+1 ));
+
+* Suggest Quote the right-hand side of == in [[ ]] to prevent glob matching, the left hand side can be quoted too.
+```
+run bash hamming.sh 'AAA' 'A?A'
+
+✗ expose subtle '[[ $x == $y ]]' bug
+   (in test file hamming_test.sh, line 92)
+     `(( status == 0 ))' failed
+```
+
+Problem line of code: 
+```bash
+[[ ${a:i:1} == ${b:i:1} ]] || count+=1
+```
+Potential fix:
+```bash
+[[ "${a:i:1}" == "${b:i:1}" ]] || count+=1
+```
+
+
+### Talking Points
+
+* `if`/`else` used instead of List Constructs.
+
+Because conditionals are needed. If the
+student uses `if`/`else` it's an opportunity to talk about the list constructs for a more concise answer. `&&`/`||` The and list and or list constructs provide a means of processing a number of commands consecutively. These can effectively replace complex nested if/then or even case statements. [List Constructs](https://tldp.org/LDP/abs/html/list-cons.html#LISTCONSREF)
```

Comment:
```suggestion
student uses `if`/`else` it's an opportunity to talk about the list constructs for a more concise answer. `&&`/`||` The and list and or list constructs provide a means of processing a number of commands consecutively. These can effectively replace complex nested `if`/`then` or even case statements. [List Constructs](https://tldp.org/LDP/abs/html/list-cons.html#LISTCONSREF)
```

Reply:
I think this paragraph needs further attention.  Quote "and list" and "or list" maybe?  Not quite sure.

Reply:
This talking point has been contraversial. 

I am leaning towards removing it and it can be added back in as a future pr. 

@IsaacG , made a salient point about list comprehensions that would argue if-else should be preferred.

"
The && and || in bash are short-circuit logical operators. They quit evaluating as soon as possible.

a && b will only run b if a is true. a || b will only run b if a is false. a || b || c will stop as soon as anything is true. a && b || c will run a and ... b or c or both!! This is why a full blown if-else is often a good idea.
"

Reply:
Would the proper thing be to add both as opposing arguments

or remove both as talking points?

Reply:
I think both as talking points.  Both are valid, and the conversation is worthwhile.

Reply:
Just added the update, so that both are showing up as a talking point. 👍

Happy to remove in an additional pr if requested.